### DISCUSSION TITLE
Any way to use a local copy of the bootstrap archive?

### DISCUSSION THREAD

**Discussion by Rhys-T** [nix-community/nix-on-droid#420] [Q&A] [answered]
For some reason, I can't get to nix-on-droid.unboiled.info. The server is fine as far as the rest of the world is concerned - it's just my particular route to it that's having problems somewhere. I was able to dowlnoad the bootstrap archive by relaying it through my VPS, but now I don't know what to do to get it loaded into nix-on-droid.

What I've tried so far:
- Setting up a folder structure on my Mac that mimics the bootstrap server (file at [root]/bootstrap-release-24.05/bootstrap-aarch64.zip) and serving it using [`httplz`](https://github.com/thecoshman/http), then editing the bootstrap URL in the dialog to `http://192.168.x.y:8000/bootstrap-release-24.05`. Fails immediately, rather than waiting for a timeout like it does with the official server. `httplz` never shows any requests coming in.
- Doing the same thing with my [Rethink](https://rethinkdns.com/app) firewall turned off. Same result.
- Copying that folder structure down to my normal Termux, serving it using [`darkhttpd`](https://github.com/emikulic/darkhttpd), and setting the URL to `http://localhost:8080/bootstrap-release-24.05` (or the `127.0.0.1` version). Same result: immediate error, no requests logged by the server.
- Sticking the folder into /sdcard/Download, and entering `file:///sdcard/Download/nix-on-droid/bootstrap-release-24.05` to try and use it straight from storage. Same result.
- Long-pressing the Nix app icon and trying to go straight to a failsafe session, so I can unpack things manually. The app still insists on making me go through the bootstrap process.

Is there any way to get nix-on-droid to take my copy of this file?

(nix-on-droid is installed from F-Droid, if it matters. But I can't connect to the official server even from my Mac, so that problem isn't on the tablet end.)

**Comment by t184256** [comment, marked as answer]
I would say "double check that you don't have a dead proxy configured", but local bootstrap with file:/// failing is doubly curious. CI emulator actually bootstraps from file, namely from `file:///data/local/tmp/n-o-d/` where the zipball gets adb push'ed there beforehand. So, that's tested. Maybe the app needs to first be granted storage permissions through Settings - Apps - Nix-on-Droid - ...?

**Reply by Rhys-T** [reply]
Ah! Yeah, it didn't ask for the storage permission, but granting it manually and using the file:/// URL worked. Thanks. I should have thought of that. Frankly I didn't even know whether it would take non-HTTP(S) URLs at all - I was just trying that out of desperation.

The problem connecting to the official server is definitely a real network issue. I can traceroute to the server as far as `giga-hostin.ear2.dusseldorf1.level3.net` (62.67.20.82) before it fails. Still not sure why the local servers weren't working, though.

It's taking a long time to evaluate the initial config derivation, but I'll give it a while longer before I decide it's _really_ stuck.

**Reply by t184256** [reply]
Dead proxy theory was supposed to explain why you can't connect to localhost.

**Reply by Rhys-T** [reply]
Gotcha. I just checked, and there's no proxy set up. I do have the Rethink Firewall + DNS app running, which pretends to be a VPN app to do its job, but I had tried with it disabled too (plus I can get to both my Mac and localhost from other apps).

Also, it finally finished evaluating the config and got to a prompt.

Thanks again for your help!
